pkg/execution/executor/executor.go

```diff
@@ -1103,8 +1103,13 @@ func (e *executor) HandleResponse(ctx context.Context, i *runInstance, resp *sta
 	if len(resp.Generator) > 0 {
 		// Handle generator responses then return.
 		if serr := e.HandleGeneratorResponse(ctx, i, resp); serr != nil {
+			var compileError *expressions.CompileError
+
 			// If this is an error compiling async expressions, fail the function.
-			if strings.Contains(serr.Error(), "error compiling expression") {
+			if errors.As(serr, &compileError) {
+				resp.Output = nil
+				resp.Err = compileError.Message()
+
```

Comment:
I suspect we may need to also add the onFunctionFailed lifecycle hook just below the finalize call, too, actually - unless this is already handled when returning an error from the handleResponse call?

Without this lifecycle hook we won't be properly storing logs/traces/etc.

It's actually a little annoying that we have to call the lifecycle hooks _and_ finalize separately;  maybe in the future we can refactor the executor code to combine these.

Reply:
We do use the `OnFunctionFinished` one, this might be the wrong hook then? It does already handle some errors (step limit exceeded), though.